prime-router/src/main/kotlin/Metadata.kt

```diff
@@ -111,46 +125,93 @@ object Metadata {
-        this.schemas = fixupSchemas(schemas).map { normalizeSchemaName(it.name) to it }.toMap()
-    }
-
     private fun fixupSchemas(schemas: List<Schema>): List<Schema> {
         return schemas.map { schema ->
-            if (schema.extends == null && schema.basedOn == null) return@map schema
-
             // find the base schema
-            val baseSchema = schemas.find {
-                val basedOnSchema = schema.extends ?: schema.basedOn ?: ""
-                normalizeSchemaName(it.name) == normalizeSchemaName(basedOnSchema)
-            } ?: error("'basedOn/extends schema does not exist for '${schema.name}'")
+            val extendsSchema = schema.extends?.let { extendSchemaName ->
+                schemas.find {
+                    normalizeSchemaName(it.name) == normalizeSchemaName(extendSchemaName)
+                } ?: error("extends schema does not exist for '${schema.name}'")
+            }
+            val basedOnSchema = schema.basedOn?.let { basedOnSchemaName ->
+                schemas.find {
+                    normalizeSchemaName(it.name) == normalizeSchemaName(basedOnSchemaName)
+                } ?: error("basedOn schema does not exist for '${schema.name}'")
+            }
+            val baseSchema = extendsSchema ?: basedOnSchema
 
             // Inherit properties from the base schema
             var schemaElements = schema.elements.map { element ->
-                baseSchema.findElement(element.name)
-                    ?.let { element.inheritFrom(it) }
-                    ?: element
+                baseSchema?.findElement(element.name)
+                    ?.let { fixupElement(element, it) }
+                    ?: fixupElement(element)
             }
 
             // Extend the schema if there is a extend element
-            schema.extends?.let {
-                val baseElements = baseSchema.elements.mapNotNull { element ->
-                    if (schema.containsElement(element.name)) null else element
+            extendsSchema?.let {
+                val extendElements = it.elements.mapNotNull { element ->
+                    if (schema.containsElement(element.name)) null else fixupElement(element)
                 }
-                schemaElements = schemaElements.plus(baseElements)
+                schemaElements = schemaElements.plus(extendElements)
             }
-            schema.copy(elements = schemaElements)
+            schema.copy(elements = schemaElements, basedOnRef = basedOnSchema, extendsRef = extendsSchema)
         }
     }
 
-    fun findSchema(name: String): Schema? {
-        return schemas[normalizeSchemaName(name)]
-    }
-
     private fun normalizeSchemaName(name: String): String {
         return name.toLowerCase()
     }
 
+    /**
+     * The fixup process fills in references and inherited attributes.
+     */
+    private fun fixupElement(element: Element, baseElement: Element? = null): Element {
+        if (element.canBeBlank == true && element.default != null)
```

Comment:
Element fixup now fixes up references to mappers, tables, and valuesets. Code which use elements is simpler and faster.

Reply:
Just one level for now, but I think we will need multiple levels as I do the Redox schemas.